mllam_data_prep/derived_variables.py

```diff
@@ -0,0 +1,478 @@
+import datetime
+import importlib
+import sys
+
+import numpy as np
+import xarray as xr
+from loguru import logger
+
+
+def derive_variables(ds, derived_variables, chunking):
+    """
+    Load the dataset, and derive the specified variables
+
+    Parameters
+    ---------
+    ds : xr.Dataset
+        Source dataset
+    derived_variables : Dict[str, DerivedVariable]
+        Dictionary with the variables to derive with keys as the variable
+        names and values with entries for kwargs and function to use in
+        the calculation
+    chunking: Dict[str, int]
+        Dictionary with keys as the dimensions to chunk along and values
+        with the chunk size
+
+    Returns
+    -------
+    xr.Dataset
+        Dataset with derived variables included
+    """
+
+    ds_derived_vars = xr.Dataset()
+    ds_derived_vars.attrs.update(ds.attrs)
+    for _, derived_variable in derived_variables.items():
+        required_kwargs = derived_variable.kwargs
+        function_name = derived_variable.function
+        derived_variable_attributes = derived_variable.attributes
+
+        # Separate the lat,lon from the required variables as these will be derived separately
+        logger.warning(
+            "Assuming that the lat/lon coordinates are given as variables called"
+            " 'lat' and 'lon'."
+        )
+        latlon_coords_to_include = {}
+        for key in list(required_kwargs.keys()):
+            if key in ["lat", "lon"]:
+                latlon_coords_to_include[key] = required_kwargs.pop(key)
+
+        # Get input dataset for calculating derived variables
+        ds_input = ds[required_kwargs.keys()]
+
+        # Any coordinates needed for the derivation, for which chunking should be performed,
+        # should be converted to variables since it is not possible for *indexed* coordinates
+        # to be chunked dask arrays
+        chunks = {
+            dim: chunking.get(dim, int(ds_input[dim].count())) for dim in ds_input.dims
+        }
+        required_coordinates = [
+            req_var for req_var in required_kwargs.keys() if req_var in ds_input.coords
+        ]
+        ds_input = ds_input.drop_indexes(required_coordinates, errors="ignore")
+        for req_coord in required_coordinates:
+            if req_coord in chunks:
+                ds_input = ds_input.reset_coords(req_coord)
+
+        # Chunk the dataset
+        ds_input = _chunk_dataset(ds_input, chunks)
+
+        # Add function arguments to kwargs
+        kwargs = {}
+        if len(latlon_coords_to_include):
+            latlon = get_latlon_coords_for_input(ds)
+            for key, val in latlon_coords_to_include.items():
+                kwargs[val] = latlon[key]
+        kwargs.update({val: ds_input[key] for key, val in required_kwargs.items()})
+        func = _get_derived_variable_function(function_name)
+        # Calculate the derived variable
+        derived_field = func(**kwargs)
```

Comment:
Currently I have the check after `derived_field = func(**kwargs)` where I check on if `field` is an `xr.DataArray` or `Tuple(xr.DataArray)` (since currently the time forcings are outputted as tuples since they have both cos and sin components).

I have, however, not added any functionality for turning `field` into an `xr.DataArray`. Should this be added perhaps?

Reply:
I think that would be good to have too